### PR TITLE
prov/efa: Add initial unit test implementation

### PR DESCRIPTION
**Pull request by wckzhang**
Signed-off-by: William Zhang <wilzhang@amazon.com>

### CODE REVIEW

File: prov/efa/test/README.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# EFA unit tests
+
+## How to run
+
+To run efa unit tests, you will need to have cmocka installed.
+See: https://cmocka.org/
+
+You will need to configure libfabric with ```--enable-efa-unit-test```.
```
Comment (shijin-aws) [inline prov/efa/test/README.md:8]: A high level question: Does this unit test require efa device? If I configure libfabric with such flag and `make check` on a node that does not have efa device, will the test fail?

**Comment by wckzhang** [comment]
The appveyor failure looks legitimate though I'm not sure why it can't find the header file

**Comment by wckzhang** [comment]
Added header check for cmocka.h and cleaned up the autotools code a bit

**Comment by wckzhang** [comment]
Looks like I screwed something up while cleaning up the autotools, let me fix that

**Comment by wckzhang** [comment]
Should be fixed now

File: prov/efa/src/efa_static_headers.h [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+#include "efa.h"
+#ifdef EFA_UNIT_TEST
+
+/**
+ * This file contains headers for static functions that are converted
+ * to non static functions when compiled under unit testing.
+ *
+ */
+
+int rxr_info_to_rxr(uint32_t version, const struct fi_info *core_info,
```
Comment (wzamazon) [inline prov/efa/src/efa_static_headers.h:15]: Instead of testing `rxr_info_to_rxr`, I think `rxr_getinfo`  is a better target, which is the entry point of application, and  the input and output is well defined. (or even `rxr_prov->get_info`, if testing a struct is possible)

File: prov/efa/src/rxr/rxr_ep.c [hunk 3]
```diff
@@ -1896,7 +1896,7 @@ static inline void rxr_ep_check_peer_backoff_timer(struct rxr_ep *ep)
  * @param[in]	ep		end point
  * @param[in]	cqe_to_process	max number of cq entry to poll and process
  */
-static inline void rdm_ep_poll_ibv_cq(struct rxr_ep *ep,
+unit_static inline void rdm_ep_poll_ibv_cq(struct rxr_ep *ep,
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:1899]: if we are not testing `rdm_ep_poll_ibv_cq` then we do not need this change.

File: prov/efa/test/efa_test_getinfo.c [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+/* Currently unused but useful to fake for future unit tests */
+int __real_rxr_get_lower_rdm_info(uint32_t version, const char *node,
+                           const char *service, uint64_t flags,
+                           const struct util_prov *util_prov,
+                           const struct fi_info *util_hints,
+                           struct fi_info **core_info);
+
+int __wrap_rxr_get_lower_rdm_info(uint32_t version, const char *node,
```
Comment (wzamazon) [inline prov/efa/test/efa_test_getinfo.c:25]: Instead of mocking `rxr_get_lower_rdm_info`. We need to mock functions like `ibv_open_device`.  `ibv_query_device`, `ibv_query_gid`. Because the behavior is well defined.

File: prov/efa/test/efa_test_getinfo.c [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	int ret;
+	struct fi_info *core_info = fi_allocinfo();
+	struct fi_info *info = fi_allocinfo();
+	struct fi_info *hints = fi_allocinfo();
+
+	will_return(__wrap_rxr_info_to_rxr, true);
+	ret = rxr_info_to_rxr(0, core_info, info, hints);
+	assert_int_equal(ret, 0);
+	assert_int_equal(info->caps, rxr_info.caps);
+	assert_int_equal(info->mode, rxr_info.mode);
```
Comment (wzamazon) [inline prov/efa/test/efa_test_getinfo.c:67]: The first 2 tests does not do much, and can be ignored. I would really like to see a test case that we use mock function to return multiple EFA devices.

File: Makefile.am [hunk 6]
```diff
@@ -15,6 +15,7 @@ AM_CPPFLAGS = \
 
 noinst_LTLIBRARIES =
 lib_LTLIBRARIES =
+check_PROGRAMS =
```
[the remark this answers is not in the document]
Comment (wckzhang) [reply Makefile.am:18]: can't add to check programs if not initialized

File: prov/efa/src/rxr/rxr_ep.c (hunk 3 above)
Comment (wckzhang) [reply prov/efa/src/rxr/rxr_ep.c:1899]: I can remove this change

File: prov/efa/test/efa_test_getinfo.c (hunk 4 above)
Comment (wckzhang) [reply prov/efa/test/efa_test_getinfo.c:25]: Sure

**Comment by wckzhang** [comment]
Removed static mocking until further needed. Changed testing to use getinfo tests.

Wanted to do some mocking with ibv_get_device_list, but this runs on symbol loading...which makes it impossible to manipulate as a per unit test function. We'll have to fix values as far as I can tell.

**Comment by wckzhang** [comment]
@shefty can you let me know what the intel failure is?

**Comment by shefty** [comment]
No idea what the Intel failure was, just said aborted.  Maybe the CI nodes were getting sleepy.  Restarted it.

File: prov/efa/test/efa_unit_tests.c [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+int main(void)
+{
+	int ret;
+	const struct CMUnitTest getinfo_tests[] = {
+		cmocka_unit_test(efa_getinfo_default),
+	};
+	cmocka_set_message_output(CM_OUTPUT_XML);
+
+	ret = cmocka_run_group_tests_name("efa getinfo tests", getinfo_tests, NULL, NULL);
+	if (ret != 0)
```
Comment (a-szegel) [inline prov/efa/test/efa_unit_tests.c:11]: Is there any reason for this if statement?

File: prov/efa/Makefile.include [hunk 8]
```diff
@@ -73,9 +73,26 @@ _efa_headers = \
 	prov/efa/src/rxr/rxr_atomic.h \
 	prov/efa/src/rxr/rdm_proto_v4.h
 
+
+if ENABLE_EFA_UNIT_TEST
+check_PROGRAMS += prov/efa/test/efa_unit_test
+TESTS += prov/efa/test/efa_unit_test
+nodist_prov_efa_test_efa_unit_test_SOURCES = \
```
[the remark this answers is not in the document]
Comment (a-szegel) [reply prov/efa/Makefile.include:80]: I could see us wanting multiple UT executables, if our UT's are doing things that corrupt "normal" state of the system (i.e.. impact each other), but I would rather have 1 main executable that holds all of our test with an excluded list for tests that need their own executable (vs the current implementation's include list).

File: prov/efa/test/efa_unit_tests.c (hunk 7 above)
Comment (wckzhang) [reply prov/efa/test/efa_unit_tests.c:11]: At some point I wrote an example that was supposed to be "tests that fail" and had this check, but I can remove it.

File: prov/efa/Makefile.include (hunk 8 above)
[the remark this answers is not in the document]
Comment (wckzhang) [reply prov/efa/Makefile.include:80]: I'm not sure if we can do that. We might want to run the same tests under different configurations, such as "one nic" and "four nics".

It is possible to do a wildcard and add all files in a directory, but that's generally suggested against with autotools and I'm not sure that wildcarding is entirely system agnostic. In general, file additions should be uncommon enough that manual steps, or at most a script that adds the files to a Makefile.am is all that is needed.

**Comment by wckzhang** [comment]
> Can I mock any internal function in libfabric with the current setup (common or provider code), i.e.. things that are in the executable, or are we relying on a linker trick, and only supporting mocking functions in external shared libraries?

This should be able to mock any internal function as long as you include the appropriate headers.

> We should keep all of our mocks in one common place ex: prov/efa/test/scaffolding. We should strive to develop a common mock library, which can be utilized by all unit tests. Instead of pre-developing the mock library, I think that we should make a template mock + directions, to be copied/pasted by dev's who are creating new mocks for their UT's. The template will encourage developers to build their new mocks in a common way which will drive reuse for our tests, and populate our mock library as we go.

This is doable

File: prov/efa/test/README.md (hunk 1 above)
Comment (a-szegel) [reply prov/efa/test/README.md:8]: Tests must have changed... now they do.

File: prov/efa/configure.m4 [hunk 9]
```diff
@@ -113,6 +115,35 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
...
+		[efa_unit_test=$enableval],
+		[efa_unit_test=no])
+
+	if test x"$efa_unit_test" = x"yes"; then
+		AC_CHECK_HEADERS([cmocka.h],
+			[
+				AC_DEFINE([EFA_UNIT_TEST], [1],
+					[EFA unit testing])
+				efatest_CPPFLAGS="$efatest_CPPFLAGS $efadv_CPPFLAGS"
+				efatest_LIBS="-lcmocka -libverbs -static $efatest_LIBS"
```
Comment (wzamazon) [inline prov/efa/configure.m4:131]: Do not you need find cmocka and setup include path?

File: prov/efa/Makefile.include [hunk 10]
```diff
@@ -73,9 +73,26 @@ _efa_headers = \
...
+	prov/efa/test/efa_unit_tests.h \
+	prov/efa/test/efa_unit_tests.c \
+	prov/efa/test/efa_test_getinfo.c
+
+prov_efa_test_efa_unit_test_LDADD = $(efatest_LIBS) $(linkback)
+prov_efa_test_efa_unit_test_CPPFLAGS = -I$(top_srcdir)/include -pthread $(efatest_CPPFLAGS) -I$(top_srcdir)/prov/efa/src -I$(top_srcdir)/prov/efa/src/rxr -I$(top_srcdir)/prov/efa/test
+prov_efa_test_efa_unit_test_LDFLAGS = $(efatest_LDFLAGS) -Wl,--wrap=rxr_info_to_rxr,--wrap=ibv_get_device_list
```
Comment (wzamazon) [inline prov/efa/Makefile.include:87]: `rxr_info_to_rxr` is not needed any more. right?

Comment (wckzhang) [reply prov/efa/Makefile.include:87]: Yeah let me remove that

File: prov/efa/Makefile.include [hunk 11]
```diff
@@ -73,9 +73,26 @@ _efa_headers = \
 	prov/efa/src/rxr/rxr_atomic.h \
 	prov/efa/src/rxr/rdm_proto_v4.h
 
+
+if ENABLE_EFA_UNIT_TEST
```
Comment (a-szegel) [inline prov/efa/Makefile.include:77]: When I configure with for unit testing, I still need to manually add the cmocka library path to my LD_LIBRARY_PATH for `make check` to work.  The output of `make check` is dense and hard to parse visually, and I think the tester needs to `cat prov/efa/test/efa_unit_test.log` to see what went wrong.

File: prov/efa/test/README.md [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+simpler, you can add a macro "unit_static" that will change static functions to
+non static when --enable-efa-unit-test is enabled. You will also need to add
+a new declaration by creating a file such as "efa_static_headers.h".
+
+### Manipulating mock behavior
+
+You can use the cmocka API to change the behavior of your mock function.
+
+The will_return class of functions will place values onto a stack that can
+be popped within the function with the mock function. You can use this to
```
Comment (mmubarak) [inline prov/efa/test/README.md:44]: @wckzhang added a few comments regarding README in this PR: https://github.com/ofiwg/libfabric/pull/7472

File: prov/efa/test/README.md (hunk 1 above)
Comment (a-szegel) [reply prov/efa/test/README.md:8]: Currently, an EFA device is required.

File: prov/efa/src/efa.h [hunk 13]
```diff
@@ -111,6 +114,12 @@
 /* maximum name length for shm endpoint */
 #define EFA_SHM_NAME_MAX	   (256)
 
+#ifdef EFA_UNIT_TEST
+#define unit_static
+#else
+#define unit_static static
```
[the remark this answers is not in the document]
Comment (a-szegel) [reply prov/efa/src/efa.h]: OBE, Team decided to put the UT's in the same file as the src, so we can unit test static functions now.

File: prov/efa/test/efa_test_getinfo.c (hunk 5 above)
Comment (a-szegel) [reply prov/efa/test/efa_test_getinfo.c]: @wzamazon Is this still blocking the PR from being merged?

File: prov/efa/Makefile.include (hunk 11 above)
Comment (a-szegel) [reply prov/efa/Makefile.include:77]: Fixing this with --cmocka <path>, and adding that to rpath.

File: prov/efa/Makefile.include [hunk 14]
```diff
@@ -73,9 +73,26 @@ _efa_headers = \
...
+check_PROGRAMS += prov/efa/test/efa_unit_test
```
[the remark this answers is not in the document]
Comment (a-szegel) [reply prov/efa/Makefile.include:78]: Talked with Wei, he would rather run `make check` on a compute node in the cluster, or manually attach a EFA Device to the head of the cluster.

**Comment by wckzhang** [comment]
Continuing this PR in PR #7472
